### ISSUE TITLE
Conflict with spring-data-mongodb version 2.2.0.M2

### ISSUE THREAD

**Issue by jesse1993** [AxonFramework/extension-mongo#5]
Spring-data 2.2.0.M2 support `ReactiveQuerydslPredicateExecutor`, that is exciting!!!
Unfortunately, I have encountered a conflict.
Hope this issue will be solved and they can work well together.

```
java.lang.NoClassDefFoundError: com/mongodb/client/ClientSession
	at java.lang.Class.getDeclaredMethods0(Native Method) ~[na:1.8.0_121]
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701) ~[na:1.8.0_121]
	at java.lang.Class.getDeclaredMethods(Class.java:1975) ~[na:1.8.0_121]
	at org.axonframework.common.ReflectionUtils.methodsOf(ReflectionUtils.java:229) ~[axon-messaging-4.1.jar:4.1]
	at org.axonframework.spring.config.SpringAxonAutoConfigurer.lambda$registerEventHandlerRegistrar$42(SpringAxonAutoConfigurer.java:254) ~[axon-spring-4.1.jar:4.1]
	at java.util.Iterator.forEachRemaining(Iterator.java:116) ~[na:1.8.0_121]
	at org.axonframework.spring.config.SpringAxonAutoConfigurer.registerEventHandlerRegistrar(SpringAxonAutoConfigurer.java:248) ~[axon-spring-4.1.jar:4.1]
	at org.axonframework.spring.config.SpringAxonAutoConfigurer.registerBeanDefinitions(SpringAxonAutoConfigurer.java:214) ~[axon-spring-4.1.jar:4.1]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.lambda$loadBeanDefinitionsFromRegistrars$1(ConfigurationClassBeanDefinitionReader.java:358) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at java.util.LinkedHashMap.forEach(LinkedHashMap.java:684) ~[na:1.8.0_121]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsFromRegistrars(ConfigurationClassBeanDefinitionReader.java:357) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:145) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:117) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:328) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:233) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:271) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:91) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:694) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:532) ~[spring-context-5.0.9.RELEASE.jar:5.0.9.RELEASE]
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:61) ~[spring-boot-2.0.5.RELEASE.jar:2.0.5.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:780) [spring-boot-2.0.5.RELEASE.jar:2.0.5.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:412) [spring-boot-2.0.5.RELEASE.jar:2.0.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:333) [spring-boot-2.0.5.RELEASE.jar:2.0.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1277) [spring-boot-2.0.5.RELEASE.jar:2.0.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1265) [spring-boot-2.0.5.RELEASE.jar:2.0.5.RELEASE]
	at o.g.autumn.user.ApplicationQ.main(ApplicationQ.java:16) [classes/:na]
Caused by: java.lang.ClassNotFoundException: com.mongodb.client.ClientSession
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381) ~[na:1.8.0_121]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[na:1.8.0_121]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331) ~[na:1.8.0_121]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[na:1.8.0_121]
	... 26 common frames omitted
```

**Comment by smcvb** [comment]
So, @jesse1993, are you looking for a dependency upgrade of MongoDB for this extension?
I am not completely sure what you're request is from this issue, hence why I'm asking.

**Comment by jesse1993** [comment]
Yes, i am looking for a dependency upgrade of MongoDB for this extension.

**Comment by smcvb** [comment]
Thanks for the update on this part @jesse1993, that makes your request clearer.
We're currently hard at work in other areas of the Axon software suite, thus I'd argue the priority from AxonIQ's stance on this is quite low.

Additionally, we are not using the Spring Data MongoDB dependency for the Mongo extension, but are using Mongo directly. The idea behind this is to not pull in a ton of dependencies/libraries which deviate away from a pure, concise solution. 

So, when I am talking about a dependency upgrade for this extension, I'd expect to update Mongo directly. I am not anticipating this repository will include the `spring-data-mongodb` directly.

**Comment by mapkuff** [comment]
https://github.com/spring-projects/spring-boot/blob/v2.3.0.RELEASE/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoAutoConfiguration.java

FYI since spring boot 2.3.0.Release
Spring use `com.mongodb.client.MongoClient` instead of `com.mongodb.MongoClient`

and one more quick question
Do you plan to support `com.mongodb.reactivestreams.client.MongoClient` on axon `MongoTemplate` ?

**Comment by smcvb** [comment]
> Do you plan to support com.mongodb.reactivestreams.client.MongoClient on axon MongoTemplate ?

At the moment we do not have any plans around this topic, or any other major adjustments to the Mongo Extension for that matter. Contributions are always welcome of course, but as it stands the AxonIQ team is prioritized on other subjects first.

I am confident however we will revisit all the extensions, including this one, as soon as time allows it.

**Comment by smcvb** [comment]
I actually believe this original issue has been resolved already by @bjornharvold in #11. The original pain point of the package switch of the `MongoClient` is at least now part of the current Mongo implementations, which will be part of release 4.4.

Hence, I will be closing this issue as resolved. 
Feel free to keep commenting under this issue if necessary.
